google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITInlineCommitDatabaseClientTest.java

```diff
@@ -0,0 +1,166 @@
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.SpannerException;
+import com.google.cloud.spanner.Statement;
+import com.google.cloud.spanner.TimestampBound;
+import java.util.Collections;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Integration tests for InlineCommit used in DatabaseClient. */
+@Category(ParallelIntegrationTest.class)
+@RunWith(JUnit4.class)
+public final class ITInlineCommitDatabaseClientTest {
+  @ClassRule public static IntegrationTestEnv env = new IntegrationTestEnv();
+  private static DatabaseClient client;
+
+  /** Id prefix per test case. */
+  private static int id;
+
+  private static final String INSERT_DML =
+      "INSERT INTO T (k, v) VALUES ('%d-boo1', 1), ('%d-boo2', 2), ('%d-boo3', 3), ('%d-boo4', 4);";
+  private static final String COUNT = "SELECT COUNT(*) FROM T";
+  private static final long DML_COUNT = 4;
+
+  @BeforeClass
+  public static void setUpDatabase() {
+    Database db =
+        env.getTestHelper()
+            .createTestDatabase(
+                "CREATE TABLE T ("
+                    + "  K    STRING(MAX) NOT NULL,"
+                    + "  V    INT64,"
+                    + ") PRIMARY KEY (K)");
+    client = env.getTestHelper().getDatabaseClient(db);
+  }
+
+  @Before
+  public void increaseTestIdAndDeleteTestData() {
+    client.writeAtLeastOnce(Collections.singletonList(Mutation.delete("T", KeySet.all())));
+    id++;
+  }
+
+  private String insertDml() {
+    return String.format(INSERT_DML, id, id, id, id);
+  }
+
+  // Verifies that one of the four new records (in particular with key=$id-boo1) should exist.
+  private void verifyInserted() {
+    assertEquals(
+        1,
+        client
+            .singleUse(TimestampBound.strong())
+            .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
+            .getLong(0));
+  }
+
+  @Test
+  public void inlineCommit() {
+    client
+        .readWriteTransaction(Options.inlineCommitOption())
+        .run(
+            transaction -> {
+              long rowCount = transaction.executeUpdate(Statement.of(insertDml()));
+              assertEquals(DML_COUNT, rowCount);
+              return null;
+            });
+    verifyInserted();
+  }
+
+  @Test
+  public void queryThenInlineCommit() {
+    client
+        .readWriteTransaction(Options.inlineCommitOption())
+        .run(
+            transaction -> {
+              // Do a query.
+              ResultSet resultSet = transaction.executeQuery(Statement.of(COUNT));
+              assertTrue(resultSet.next());
+              assertEquals(0, resultSet.getLong(0));
+              assertFalse(resultSet.next());
+
+              // DML after query should succeed and should cause the transaction to be committed.
```

Comment:
I think that this will be a confusing API for users. The transaction is in `autocommit` mode, which means that you can execute any number of queries, but the first DML statement that is executed will end the transaction. I understand why it technically works like that, but I do worry that this might cause confusion.

Reply:
One thing we may do is to ensure the comments at Options.inlineCommitOption is as clear as possible; e.g. we may say setting this requires the transaction to have nothing but a single DML statement. Are there better options?

Reply:
See my comments above. I think that if this is really only intended for executing a single DML statement, that we should implement it as a single method that only accepts one DML statement. That could be either something like:

```java
DatabaseClient client = spanner.getDatabaseClient(..);
client.executeUpdate("UPDATE Foo");
```

or something like

```java
DatabaseClient client = spanner.getDatabaseClient(..);
client.singleUseTransaction().executeUpdate("UPDATE Foo");
```

Reply:
removed Options.inlineCommitOption.